[v3 cog manager, utils] handle missing cogs correctly, add some helpful algorithms

calebj opened the pull request:
### Type

- [x] Bugfix
- [x] Enhancement
- [x] New feature

### Description of the changes
For cog loading, only show "cog not found" if the module in question was the one that failed to import. `ImportErrors` within cogs will now show an error as they should.

Also:
- Add a chain deduplicator, benchmarked to be the fastest method
- Add a semaphore-bounded variant of `asyncio.gather` and an async `as_completed`
- And of course, tests and docs for all additions

calebj committed:
Handle missing cogs correctly, add some helpful algorithms

For cog loading, only show "cog not found" if the module in question was the one
that failed to import. ImportErrors within cogs will show an error as they should.

- deduplicator, benchmarked to be the fastest
- bounded gather and bounded async as_completed
- tests for all additions

calebj commented:
@Tobotimus rebased.

Tobotimus commented on redbot/core/utils/__init__.py:
```diff
@@ -70,10 +95,105 @@ def safe_delete(pth: Path):
+        loop = asyncio.get_event_loop()
+
+    if semaphore is None:
+        if type(limit) != int or limit <= 0:
```
Please use `not isinstance` instead of `type() !=`

Tobotimus commented on redbot/core/utils/__init__.py:
```diff
@@ -70,10 +95,105 @@ def safe_delete(pth: Path):
+            try:
+                done, pending = await asyncio.wait(pending, loop=loop, return_when=FIRST_COMPLETED)
+            except asyncio.CancelledError:
+                []
```
Not sure what's going on with this `[]` statement, would prefer the use of `pass` or `contextlib.supress`

calebj replied:
I think I had logic here for cancelling pending tasks, but I found something better (see the push)

Tobotimus commented on redbot/core/utils/__init__.py:
```diff
@@ -70,10 +95,105 @@ def safe_delete(pth: Path):
+
+    async def sem_wrapper(sem, task):
+        async with sem:
+            return await task
```
Since this `sem_wrapper` is common to the one in `bounded_gather_iter`, it can be moved to the module-level and made private

Tobotimus commented on redbot/core/utils/__init__.py:
```diff
@@ -70,10 +95,105 @@ def safe_delete(pth: Path):
+        loop = asyncio.get_event_loop()
+
+    if semaphore is None:
+        if type(limit) != int or limit <= 0:
```
`isinstance` here too

Tobotimus commented on redbot/core/core_commands.py:
```diff
@@ -23,6 +23,7 @@
 from redbot.core import checks
 from redbot.core import i18n
 from redbot.core import commands
+from redbot.core.cog_manager import NoSuchCog
```
Unused import

Tobotimus commented on redbot/core/cog_manager.py:
```diff
@@ -209,10 +211,10 @@ def _ensure_path_obj(path: Union[Path, str]) -> Path:
+            When no cog with the requested name was found.
         """
-        resolved_paths = _deduplicate(await self.paths())
+        resolved_paths = deduplicate_iterables(await self.paths())
```
`paths()` already deduplicates the paths before it returns, so you can remove the call to deduplicate here

Tobotimus reviewed (changes requested):
@calebj This PR adds a lot of really nice things, thank you. Just a few little things that can be fixed up and then I think it's good to go.

calebj committed:
Requested changes + wrap as_completed instead

So I went source diving and realized as_completed works the way I want it to,
and I don't need to reinvent the wheel for cancelling tasks that remain
if the generator is `break`ed out of. So there's that.

mikeshardmind reviewed (approved):
I'm not seeing any issues with this after the changes made. I'm giving it a :+1:

Tobotimus reviewed (approved):
Awesome. Thanks @calebj and thanks @mikeshardmind for reviewing
